src/misc/ansi_code_markup.cpp

```diff
@@ -0,0 +1,200 @@
+		if (close) {
+			// "closing" tags have ascii codes +20
+			ansi_num += 20 + (tag == "b" ? 1 : 0); // [/b] is the
+			                                       // same as [/dim]
+		}
+		safe_sprintf(ansi_code, "\033[%dm", ansi_num);
+		return ansi_code;
+	}
+
+private:
+	ColorParser color;
+	char ansi_code[10] = {0};
+	std::unordered_map<std::string, int> tags = {
+	        {"reset", 0}, {"b", 1},       {"dim", 2},
+	        {"i", 3},     {"u", 4},       {"s", 9},
+	        {"blink", 5}, {"inverse", 7}, {"hidden", 8},
+	};
+};
+
+static TagParser tag_parser;
+
+static std::regex markup(R"((\\)?)" // Escape tag?
+						 "(\\["      // Opening bracket, open main group
+                         "[ \\t]*?" // Optional spacing after opening bracket
+                         "(\\/)?"   // Check for losing tag
+                         "("        // Start group of tags
+                         "((?:bg)?color)" // Select color or bgcolor. bg not
+                                          // captured in separate group
+                         "(?:[ \\t]*?=[ \\t]*?([a-z\\-]+))?" // Color value. '='
+                                                             // not captured in
+                                                             // separate group.
+                                                             // Spacing around
+                                                             // '=' is allowed
+                         "|i|b|u|s|blink|dim|hidden|inverse|reset" // All other
+                                                                   // tags to match
+                         ")"        // End group of tags
+                         "[ \\t]*?" // Optional spacing before closing bracket
+                         "\\])" // Closing bracket. Note: Clang and MSVC requires
+                               // this to be escaped. Closing main group
+                         ,
+                         std::regex::icase);
+
+std::string convert_ansi_markup(std::string &str)
+{
+	return convert_ansi_markup(str.c_str());
+}
+
+/*!
+ * \brief Convert ANSI markup tags in string to ANSI terminal codes
+ *
+ * Tags are in the form of [tagname]some text[/tagname]. Not all tags
+ * have closing counterparts, such as [reset], [color], and [bgcolor].
+ *
+ * Color tags take a required parameter in the form [color=value] or
+ * [bgcolor=value]. Tag matching is case insensitive, and spacing is
+ * allowed between '[', ']' and '='.
+ *
+ * If a tag cannot be parsed for any reason, the resulting string will
+ * contain the original unparsed tag.
+ *
+ * Existing ANSI terminal codes are preserved in the output string.
+ *
+ * \param str string containing ANSI markup tags
+ * \return std::string
+ */
+std::string convert_ansi_markup(const char *str)
+{
+	std::string result;
+	const char *begin = str;
+	const char *last_match = str;
+	std::cmatch m;
+	while (std::regex_search(begin, m, markup)) {
+		const char* r = nullptr;
```

Comment:
(probably not needed), but cmatch docs mention some actions are undefined if its`ready()` function is false.  Maybe we could add `assert(m.ready());` before line 175?

Reply:
Also noticed we can check if `m.empty()`, I'm not sure if that's useful anywhere (as an early bail out / continue). Disregard otherwise!

Reply:
> (probably not needed), but cmatch docs mention some actions are undefined if itsready() function is false. Maybe we could add assert(m.ready()); before line 175?

If I'm reading the documentation correctly, regex_search always sets `m.ready()` to true.

> Also noticed we can check if m.empty(), I'm not sure if that's useful anywhere (as an early bail out / continue). Disregard otherwise!

If `regex_search` does not find a match, it returns `false`. If it does find a match, then the match won't be empty.